src/include/simeng/OS/SyscallHandler.hh

```diff
@@ -103,13 +105,106 @@ struct linux_dirent64 {
+  /** Values to set modified registers to */
+  std::vector<RegisterValue> modifiedRegisterValues;
+  /** Memory address/width pairs to modify */
+  std::vector<MemoryAccessTarget> memoryAddresses;
+  /** Values to write to memory */
+  std::vector<RegisterValue> memoryAddressValues;
+};
+
+/** The result from a handled syscall. */
+struct SyscallResult {
+  /** Whether execution should halt. */
+  bool fatal;
+  /** Id of the syscall to aid exception handler processing. */
+  uint64_t syscallId;
+  // The unique ID of the core associated with the syscall
+  uint64_t coreId;
+  /** Any changes to apply to the process state. */
+  ProcessStateChange stateChange;
+};
+
+/** A struct to hold information used as arguments to a syscall. */
+struct SyscallInfo {
+  // The ID of the syscall
+  uint64_t syscallId;
+  // The unique sequenceID of the instructions triggering the syscall
+  uint64_t seqId;
+  // The unique ID of the core associated with the syscall
+  uint64_t coreId;
+  // The unique ID of the process associated with the syscall
+  uint64_t processId;
+  // The register values used a parameters to the envoked syscall
+  RegisterValue R0;
+  RegisterValue R1;
+  RegisterValue R2;
+  RegisterValue R3;
+  RegisterValue R4;
+  RegisterValue R5;
+  // The register to return the success code to
+  Register ret;
+};
+
+/** A Linux kernel syscall emulation implementation, which mimics the responses
+   to Linux system calls. */
 class SyscallHandler {
  public:
-  /** Create a new SyscallHandler object. */
+  /** Create new SyscallHandler object. */
   SyscallHandler(
-      const std::unordered_map<uint64_t, std::shared_ptr<Process>>& processes);
+      const std::unordered_map<uint64_t, std::shared_ptr<Process>>& processes,
+      std::shared_ptr<MemoryInterface> memory,
+      std::function<void(simeng::OS::SyscallResult)> returnSyscall,
+      std::function<uint64_t()> getSystemTime);
+
+  /** Tick the syscall handler to carry out any oustanding syscalls. */
+  void tick();
+
+  /** Initialise the processing of the syscall at the front of the syscallQueue_
+   * queue. */
+  void initSyscall();
+
+  /** Add the incoming syscall to the syscallQueue_ queue for later processing.
+   */
+  void receiveSyscall(const SyscallInfo info);
+
+  /** Once the syscall is deemed complete, conclude its execution by
+   * constructing a SyscallResult and supplying it to the returnSyscall_
+   * function. */
+  void concludeSyscall(ProcessStateChange change, bool fatal = false);
+
+  /** Attempt to read a string of max length `maxLength` from address `address`
+   * into the supplied buffer, starting from character `offset`. An offset of
+   * `-1` (default) will queue a read operation for the first character.
+   *
+   * This function will repeatedly set itself as the handler for the next cycle
+   * until it either reads a null character or reaches the maximum length, at
+   * which point it will call `then`, supplying the length of the read string.
+   */
```

Comment:
Assuming the `*/` is on a new line because the line above it has maximum comment length?

Reply:
Outdated